### ISSUE TITLE
error: cannot spawn git: Invalid argument

### ISSUE THREAD

**Issue by MagicAndre1981** [git-for-windows/git#1481]
- [X] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit? I use `Git-2.16.1.2-64-bit.exe`

```
$ git --version --build-options
git version 2.16.1.windows.2
cpu: x86_64
built from commit: e78e3c8ee9c219723d60aa1bccd8348c2269b9ba
sizeof-long: 4
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

Windows 8.1 x64

```
$ cmd.exe /c ver
Microsoft Windows [Version 6.3.9600]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

C:\Users\André> type "C:\Program Files\Git\etc\install-options.txt"
Editor Option: Nano
Path Option: Cmd
SSH Option: OpenSSH
CURL Option: WinSSL
CRLF Option: CRLFCommitAsIs
Bash Terminal Option: ConHost
Performance Tweaks FSCache: Enabled
Use Credential Manager: Enabled
Enable Symlinks: Enabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

n/a

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

I get the error when using TortoiseGit to to a pull.

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
I get the error when using TortoiseGit to to a pull.
```
 - What did you expect to occur after running these commands?

Pull works fine and I get the new changes pulled.

 - What actually happened instead?

I get this error:

```
git.exe pull --progress -v --no-rebase "origin"

error: cannot spawn git: Invalid argument

git did not exit cleanly (exit code 1)
```

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

happens to all cloned repositories

**Comment by garretwilson** [comment]
I just ran into this with a fresh system with Git 2.16.1.windows.2 and TortoiseGit 2.5.7.0.

I note that someone has also [asked about this problem on Stack Overflow](https://stackoverflow.com/q/48608113/421049).

**Comment by garretwilson** [comment]
Interestingly when I run the same command TortoiseGit claims it is running from the command line, it executes successfully:

```
git.exe pull --progress -v --no-rebase "origin"
```

But still TortoiseGit shows an error:

> error: cannot spawn git: Invalid argument
> 
> git did not exit cleanly (exit code 1) (63 ms @ 2/6/2018 7:23:08 AM)

Maybe this version of Git is not returning the correct error code.

**Comment by MagicAndre1981** [comment]
thanks for the link @garretwilson . I did the downgrade to 2.160. as suggested in the SO topic and it works again.

**Comment by dscho** [comment]
Isn't this a duplicate of #1475?

**Comment by MagicAndre1981** [comment]
I run 8.1 and I don't get "Function not implemented". I have no idea if this is a duplicate. can we get a nightly build to test it?

**Comment by dscho** [comment]
Due to unfortunate *other* problems, the snapshot builds did not work recently, so there are no current snapshots.

However, what is there is a new version: v2.16.1(3), just released. It's just out of the oven, you may have to wait a little until it cools off ;-)

Seriously again, if you could try that one, and report back, that would be awesome.

**Comment by MagicAndre1981** [comment]
@dscho it is NO duplicate. I just installed `Git-2.16.1.3-64-bit.exe` and still get the same error message

**Comment by MagicAndre1981** [comment]
I did the downgrade to `Git-2.16.1-64-bit.exe` and this also fixes the issue. So there is a regression between 2.16.1 and 2.16.2

**Comment by MagicAndre1981** [comment]
@dscho 
I've tried `Git-2.16.1.3-64-bit.exe` on my Windows 10 v1709 VM and I get the same error. So this is no issue related to Windows 8.1 and no duplicate.

**Comment by dscho** [comment]
> So this is no issue related to Windows 8.1 and no duplicate.

*sigh* I *just* released a new Git for Windows version! Can't I *ever* have a good version?

</rant>

Can you see (e.g. using [Process Monitor](https://docs.microsoft.com/en-us/sysinternals/downloads/procmon)) *which* `git.exe` is called (there are a couple of 'em...) in the working vs non-working case?

**Comment by Bendor** [comment]
The same problem on Windows 10 with Git-2.16.1.2-64-bit.exe and  Git-2.16.1.3-64-bit.exe build

**Comment by Tim-bflow** [comment]
Same thing here.  I can no longer Push anything because I need to Pull first and have clients waiting on my updates...

**Comment by 0x084E** [comment]
Same problem here

``` 
git stash push
error: cannot spawn git-stash: Invalid argument
```

```
Windows 10 1709 x64
MinGit busybox 2.16.1(3) x64
TortoiseGit 2.5.7 Beta x64
```

**Comment by dscho** [comment]
> MinGit busybox 2.16.1(3) x64

Wait, are y'all talking about BusyBox-backed MinGit? Why is this the first time anybody tells me about this!

**Comment by TCHessen** [comment]
No. I had the same issue on Windows 10 without the busybox build.

**Comment by dscho** [comment]
@TCHessen could you kindly describe your setup precisely? And also verify that things still do not work correctly with *Portable*Git v2.16.1(3)?

**Comment by TCHessen** [comment]
Sure. I used TortoiseGIT 2.5.7 and 

git version 2.16.1.windows.3
cpu: x86_64
built from commit: dc364ab06b41bcd142295751cef12e2c88fdc437
sizeof-long: 4

Any other infos needed?

**Comment by dscho** [comment]
Does TortoiseGit really play a role, or does it also occur if you use Git Bash or Git CMD directly?

BTW if you want, we can head over to https://gitter.im/git-for-windows/git and discuss more interactively. I would like to get to the bottom of this, and if you can help me with that, I will be very grateful.

**Comment by TCHessen** [comment]
TortoiseGIT plays a role, but I cannot say more. On the command line there occurs no error.
C:\...\xxx>"C:\Program Files\Git\bin\git.exe" pull --progress -v --no-rebase "origin"
From https://git.sittig.local/Web/xxx
 = [up to date]      master     -> origin/master
Already up to date.

**Comment by 0x084E** [comment]
I can confirm this. ``git stash push`` is also working well when called from cmd.exe

**Comment by aimbin** [comment]
After I update git  to windows3(Git-2.16.1.3-64-bit.exe), still can not use git-tortoise to pull. But can do git remote show and git pull in command line.

**Comment by dscho** [comment]
> TortoiseGIT plays a role, but I cannot say more.

Could you use the [Process Monitor](https://docs.microsoft.com/en-us/sysinternals/downloads/procmon) to log what is happening while TortoiseGit exposes that problem, then right-click on the `git.exe` that is spawned to verify its location? It *could* be that TortoiseGit bundles its own MinGit, after all.

**Comment by dscho** [comment]
@aimbin thank you for confirming. At this point, I would like to reduce the "#metoo" chatter, though, as I already *know* that it is a big problem, and do not need to be reminded of it. What I need is more information to recreate the problem here so I can work on a fix. Thank you.

**Comment by dscho** [comment]
> I did the downgrade to 2.16.0.2 as suggested in the SO topic and it works again.

@MagicAndre1981 can you confirm or deny that downgrading to v2.16.1 works?

People, I need your help.

**Comment by TCHessen** [comment]
TortoiseGIT spawns "C:\Program Files\Git\bin\git.exe" pull --progress -v --no-rebase "origin" , I can confirm it with the ProcessMonitor.

**Comment by 0x084E** [comment]
@dscho I can confirm that v2.16.1 woks

**Comment by dscho** [comment]
> I did the downgrade to Git-2.16.1-64-bit.exe and this also fixes the issue. So there is a regression between 2.16.1 and 2.16.2

Oh, nevermind @MagicAndre1981 I should have tried to read more closely. So yes, the problem is most likely that the code to try to avoid those locking issues somehow does not like the standard handles passed in by TortoiseGit.

@TCHessen thank you for confirming. Let me build a custom installer for you, with a tentative fix.

**Comment by dscho** [comment]
Okay, here it is: https://github.com/dscho/git/releases/tag/attempt-to-work-around-1481-v1

If it does not work (or if it does, and you are willing to help me even further), can you set the config variable `config.debug1481 = true` in the repository and run the Git operation again? The debug output should help me identify the root cause.

**Comment by 0x084E** [comment]
```
git.exe stash push

error: spawn2 not even attempted
error: cannot spawn git-stash: Invalid argument
```
``config.debug1481 = true`` doesn't seem to have an effect on this output

**Comment by TCHessen** [comment]
With TortoiseGIT:

git.exe pull --progress -v --no-rebase "origin"

warning: ret: 1, restrict 1, #count: 3, err: 122
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: ret: 1, restrict 1, #count: 3, err: 122
From https://git.sittig.local/Web/ZRH_XOVIS
= [up to date]      master     -> origin/master
warning: ret: 0, restrict 1, #count: 2, err: 87
error: error: 87
error: #stdhandles: 2
error: handle #0: fffffffffffffffe (file type 0, get handle info 1: 0)
error: handle #1: 00000000000005c8 (file type 3, get handle info 1: 0)
warning: spawn2 successful, err: 6
warning: ret: 0, restrict 1, #count: 2, err: 87
error: error: 87
error: #stdhandles: 2
error: handle #0: fffffffffffffffe (file type 0, get handle info 1: 0)
error: handle #1: 00000000000005c8 (file type 3, get handle info 1: 0)
warning: spawn2 successful, err: 6
Already up to date.
warning: ret: 1, restrict 0, #count: 2, err: 0

Success (1594 ms @ 07.02.2018 12:18:15)

With cmd.exe

C:\...\ZRH_XOVIS>"C:\Program Files\Git\bin\git.exe" pull --progress -v --no-rebase "origin"
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: ret: 1, restrict 1, #count: 3, err: 122
From https://git.sittig.local/Web/ZRH_XOVIS
 = [up to date]      master     -> origin/master
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: ret: 1, restrict 1, #count: 3, err: 122
Already up to date.

```

**Comment by toouur** [comment]
Hi there, I was the one who raised the question on SO. I can confirm, that the pull worked just fine both from git bash and git cmd. Also, it seems like working with tortoise if you do it with fetch rebase (recheck it please).
I haven't downgraded on my home PC yet because it's not critical for me.

**Comment by dscho** [comment]
@0x084E thanks for testing. Can you try with `git.exe -c core.debug1481=true stash push`?



> error: handle #0: fffffffffffffffe (file type 0, get handle info 1: 0)

@TCHessen now, this is fascinating. I never knew that you could pass in such a handle. `INVALID_HANDLE_VALUE` corresponds to `ffffffffffffffff` (i.e. the unsigned version of -1), while this is -2. I wonder where that comes from.

**Comment by 0x084E** [comment]
@dscho ``git.exe -c core.debug1481=true stash push`` with cmd.exe works as expected and doesn't produce any further output.
With Tortoise I cannot add additional arguments. It produces the output above.
If tried to set debug1481 at repo and user level, but output doesn't change. Am I doing somthing wrong? The config looks like this:
```
[core]
  debug1481 = true
```

**Comment by dscho** [comment]
> git.exe -c core.debug1481=true stash push with cmd.exe works as expected and doesn't produce any further output

That is not expected. `git.exe stash` should *always* spawn something. If it does not even show the "spawn2 not even attempted" message, something is really wrong, as in: this is probably not that `git.exe` you just installed from that custom installer.

> The config looks like this:
>
> ``` 
> [core]
>   debug1481 = true
> ```

If the patched `git.exe` is running, that should indeed provide plenty of debug output.

**Comment by TCHessen** [comment]
I did exactly the same and got the debug output. I hope that it helps.

**Comment by 0x084E** [comment]
@dscho see edit, for other commands i get an output
if checked this multiple times now, with other commands i get additional output but not with ``stash``, not with ``stash push`` and not with ``stash pop``. But I can see that it spawns multiple git processes.
I've also checked multiple times that it uses always the same git install.

**Comment by dscho** [comment]
@0x084E ah, maybe the problem is that the config variable is not parsed in time. That must be the problem, definitely, yes, as `git-stash` is a Unix shell script, and at that point we do not parse the config lest we taint commands that do not even need a Git directory...

I'll have a look at the edited output.

**Comment by dscho** [comment]
@0x084E sadly, the edits did not come through automatically. I had to reload the entire page. Could I ask you a favor? Just add comments, instead of editing them... 😄 

> error: handle #0: fffffffffffffffe (file type 0, get handle info 1: 0)

Aha! That's the same symptom as in the TortoiseGit case.

What is your setup? You are not using TortoiseGit, correct? So which terminal do you use?

**Comment by 0x084E** [comment]
OK, will do next time 😄 
sorry for confusion, the output above comes from TortoiseGit. If I don't use TortoiseGit I use cmd.exe which works fine in all cases I've tried so far.

**Comment by dscho** [comment]
BTW @TCHessen @0x084E I uploaded another custom installer here: https://github.com/dscho/git/releases/tag/attempt-to-work-around-1481-v2

It should print more information on the funny handle. Could you test, please?

**Comment by dscho** [comment]
> sorry for confusion, the output above comes from TortoiseGit.

@0x084E thanks for the clarification! You probably mentioned it somewhere in the thread, and my congested sinuses prevent me from holding all the information in my working memory. Sorry about that.

**Comment by 0x084E** [comment]
output from ``git.exe fetch -v --progress "origin"`` (with TortoiseGit)
```
warning: get_osf_handle for 7 (fd_is_interactive 0 0): 0000000000000284 (0000000000000000 0000000000000000)
warning: fhin: 7, hstdInput: 0000000000000284
warning: get_osf_handle for 8 (fd_is_interactive 0 0): 0000000000000288 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 0000000000000594 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 5
warning: get_osf_handle for 9 (fd_is_interactive 0 0): 0000000000000090 (0000000000000000 0000000000000000)
warning: fhin: 9, hstdInput: 0000000000000090
warning: get_osf_handle for 10 (fd_is_interactive 0 0): 0000000000000094 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 0000000000000594 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 0
warning: get_osf_handle for 10 (fd_is_interactive 0 0): 0000000000000078 (0000000000000000 0000000000000000)
warning: fhin: 10, hstdInput: 0000000000000078
warning: get_osf_handle for 12 (fd_is_interactive 0 0): 0000000000000158 (0000000000000000 0000000000000000)
warning: get_osf_handle for 11 (fd_is_interactive 0 0): 000000000000015c (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 0
warning: get_osf_handle for 5 (fd_is_interactive 0 0): 000000000000027c (0000000000000000 0000000000000000)
warning: fhin: 5, hstdInput: 000000000000027c
warning: get_osf_handle for 6 (fd_is_interactive 0 0): 0000000000000290 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 0000000000000594 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 5
warning: get_osf_handle for 11 (fd_is_interactive 0 0): 0000000000000198 (0000000000000000 0000000000000000)
warning: fhin: 11, hstdInput: 0000000000000198
warning: get_osf_handle for 12 (fd_is_interactive 0 0): 000000000000019c (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 0000000000000594 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 0
= [up to date]          develop            -> origin/develop
warning: get_osf_handle for 0 (fd_is_interactive 0 0): fffffffffffffffe (0000000000000000 0000000000000000)
warning: fhin: 0, hstdInput: fffffffffffffffe
warning: get_osf_handle for 1 (fd_is_interactive 0 0): 0000000000000594 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 0000000000000594 (0000000000000000 0000000000000000)
warning: ret: 0, restrict 1, #count: 2, err: 87
error: error: 87
error: #stdhandles: 2
error: handle #0: fffffffffffffffe (file type 0, get handle info 1: 0)
error: handle #1: 0000000000000594 (file type 3, get handle info 1: 0)
warning: spawn2 successful, err: 0
```

**Comment by TCHessen** [comment]
git.exe pull --progress -v --no-rebase "origin"

warning: get_osf_handle for 7 (fd_is_interactive 0 0): 0000000000000274 (0000000000000000 0000000000000000)
warning: fhin: 7, hstdInput: 0000000000000274
warning: get_osf_handle for 8 (fd_is_interactive 0 0): 0000000000000278 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 0000000000000598 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 5 (fd_is_interactive 0 0): 0000000000000264 (0000000000000000 0000000000000000)
warning: fhin: 5, hstdInput: 0000000000000264
warning: get_osf_handle for 6 (fd_is_interactive 0 0): 0000000000000278 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 0000000000000598 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 7 (fd_is_interactive 0 0): 0000000000000160 (0000000000000000 0000000000000000)
warning: fhin: 7, hstdInput: 0000000000000160
warning: get_osf_handle for 8 (fd_is_interactive 0 0): 0000000000000164 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 0000000000000598 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 8 (fd_is_interactive 0 0): 0000000000000150 (0000000000000000 0000000000000000)
warning: fhin: 8, hstdInput: 0000000000000150
warning: get_osf_handle for 10 (fd_is_interactive 0 0): 0000000000000160 (0000000000000000 0000000000000000)
warning: get_osf_handle for 9 (fd_is_interactive 0 0): 0000000000000164 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 5 (fd_is_interactive 0 0): 0000000000000268 (0000000000000000 0000000000000000)
warning: fhin: 5, hstdInput: 0000000000000268
warning: get_osf_handle for 6 (fd_is_interactive 0 0): 0000000000000278 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 0000000000000598 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 9 (fd_is_interactive 0 0): 0000000000000080 (0000000000000000 0000000000000000)
warning: fhin: 9, hstdInput: 0000000000000080
warning: get_osf_handle for 10 (fd_is_interactive 0 0): 000000000000006c (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 0000000000000598 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
From https://git.yyy.local/Web/xxx_XOVIS
= [up to date]      master     -> origin/master
warning: get_osf_handle for 0 (fd_is_interactive 0 0): fffffffffffffffe (0000000000000000 0000000000000000)
warning: fhin: 0, hstdInput: fffffffffffffffe
warning: get_osf_handle for 1 (fd_is_interactive 0 0): 0000000000000598 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 0000000000000598 (0000000000000000 0000000000000000)
warning: ret: 0, restrict 1, #count: 2, err: 87
error: error: 87
error: #stdhandles: 2
error: handle #0: fffffffffffffffe (file type 0, get handle info 1: 0)
error: handle #1: 0000000000000598 (file type 3, get handle info 1: 0)
warning: spawn2 successful, err: 6
warning: get_osf_handle for 0 (fd_is_interactive 0 0): fffffffffffffffe (0000000000000000 0000000000000000)
warning: fhin: 0, hstdInput: fffffffffffffffe
warning: get_osf_handle for 1 (fd_is_interactive 0 0): 0000000000000598 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 0000000000000598 (0000000000000000 0000000000000000)
warning: ret: 0, restrict 1, #count: 2, err: 87
error: error: 87
error: #stdhandles: 2
error: handle #0: fffffffffffffffe (file type 0, get handle info 1: 0)
error: handle #1: 0000000000000598 (file type 3, get handle info 1: 0)
warning: spawn2 successful, err: 6
Already up to date.
warning: get_osf_handle for 0 (fd_is_interactive 0 0): fffffffffffffffe (0000000000000000 0000000000000000)
warning: fhin: 0, hstdInput: fffffffffffffffe
warning: get_osf_handle for 1 (fd_is_interactive 0 0): 0000000000000598 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 0000000000000598 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 0, #count: 2, err: 0

Success (1812 ms @ 07.02.2018 13:19:31)

**Comment by MagicAndre1981** [comment]
here is my output of v2 of your debug exe with `debug1481 = true` set in config:

```
$  git --version --build-options
git version 2.16.1.windows.2.9.g0922cb34ea0
cpu: x86_64
built from commit: 0922cb34ea0aad80997bb8bb3df4ac6f143d6c40
sizeof-long: 4
```

```
git.exe pull --progress -v --no-rebase "origin"

warning: get_osf_handle for 8 (fd_is_interactive 0 0): 0000000000000148 (0000000000000000 0000000000000000)
warning: fhin: 8, hstdInput: 0000000000000148
warning: get_osf_handle for 9 (fd_is_interactive 0 0): 000000000000014c (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 9 (fd_is_interactive 0 0): 0000000000000138 (0000000000000000 0000000000000000)
warning: fhin: 9, hstdInput: 0000000000000138
warning: get_osf_handle for 11 (fd_is_interactive 0 0): 0000000000000148 (0000000000000000 0000000000000000)
warning: get_osf_handle for 10 (fd_is_interactive 0 0): 000000000000014c (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 10 (fd_is_interactive 0 0): 000000000000014c (0000000000000000 0000000000000000)
warning: fhin: 10, hstdInput: 000000000000014c
warning: get_osf_handle for 11 (fd_is_interactive 0 0): 0000000000000148 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
From https://github.com/telerik/xamarin-sdk
= [up to date]      master     -> origin/master
warning: get_osf_handle for 0 (fd_is_interactive 0 0): fffffffffffffffe (0000000000000000 0000000000000000)
warning: fhin: 0, hstdInput: fffffffffffffffe
warning: get_osf_handle for 1 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: ret: 0, restrict 1, #count: 2, err: 87
error: error: 87
error: #stdhandles: 2
error: handle #0: fffffffffffffffe (file type 0, get handle info 1: 0)
error: handle #1: 000000000000053c (file type 3, get handle info 1: 0)
warning: spawn2 successful, err: 6
warning: get_osf_handle for 0 (fd_is_interactive 0 0): fffffffffffffffe (0000000000000000 0000000000000000)
warning: fhin: 0, hstdInput: fffffffffffffffe
warning: get_osf_handle for 1 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: ret: 0, restrict 1, #count: 2, err: 87
error: error: 87
error: #stdhandles: 2
error: handle #0: fffffffffffffffe (file type 0, get handle info 1: 0)
error: handle #1: 000000000000053c (file type 3, get handle info 1: 0)
warning: spawn2 successful, err: 0
Already up to date.
warning: get_osf_handle for 0 (fd_is_interactive 0 0): fffffffffffffffe (0000000000000000 0000000000000000)
warning: fhin: 0, hstdInput: fffffffffffffffe
warning: get_osf_handle for 1 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: ret: 1, restrict 0, #count: 2, err: 0


Success (1516 ms @ 07.02.2018 13:32:38)
```

**Comment by dscho** [comment]
Thanks. This is getting really weird. So something either in Git or in TortoiseGit sets the handle to `(HANDLE)-2`... Let's see whether I can find some code that might do that.

**Comment by MagicAndre1981** [comment]
if you can build a version with generated PDBs and provide the PDBs for that version I can use ETW to do handle tracing with callstack so I could maybe see some details in Windows Performance Analyzer.

**Comment by dscho** [comment]
@MagicAndre1981 good idea. It is not *quite* straight-forward to do that yet, but I'll use a work-around.

In the meantime, I dug out the source code in TortoiseGit that *seems* to be responsible for calling `git.exe`: https://github.com/TortoiseGit/TortoiseGit/blob/b2d00f8785e2b148e5555925a3be7fbebca7a29f/src/Git/Git.cpp#L281-L375

As far as I can see, `hStdInput` is initialized to `0`, and never set again: https://github.com/TortoiseGit/TortoiseGit/blob/b2d00f8785e2b148e5555925a3be7fbebca7a29f/src/Git/Git.cpp#L305

However, the documentation for `STARTF_USESTDHANDLES` in [`STARTUPINFO`'s documentation](https://msdn.microsoft.com/en-us/library/windows/desktop/ms686331(v=vs.85).aspx) says this:

> If this flag is specified when calling the GetStartupInfo function, these members are either the handle value specified during process creation or `INVALID_HANDLE_VALUE`.

... which suggests to me that `hStdInput` should either be set to a valid `HANDLE` or to `INVALID_HANDLE_VALUE` (which is `-1`, not `0`). So maybe it comes from there?

@MagicAndre1981 let me get that custom build with PDBs to you. Note, however, that those are custom PDBs, as Git for Windows is built using GCC (which does not output PDBs) and the PDB files are generated using cv2pdb. So hopefully they will have enough information in them for you to do your magic.

**Comment by TCHessen** [comment]
But since this issue occurs in version 2.16.2 and 2.16.3 and not in 2.16.1 a diff should help to find the bug?

**Comment by cs96and** [comment]
I suspect it's related to [this bug fix](https://github.com/git-for-windows/git/commit/576ff26eeca22526b7ba11444da24d31daf0b369) in 2.16.1.2...

> When Git spawns processes, now only the necessary file handles are inherited from the parent process, possibly preventing file locking issues.

**Comment by dscho** [comment]
@TCHessen the bug is somewhere else than in the diff. I can show you the responsible code: https://github.com/git-for-windows/git/blob/dc364ab06b41bcd142295751cef12e2c88fdc437/compat/mingw.c#L1724-L1738

This code was introduced in v2.16.1(2), with the intention to avoid file locking issues. The thing is: if you spawn a new process, stating that you want it to inherit file handles, by default it inherits *all* of them. Not only of stdin/stdout/stderr. But also e.g. file handles to Git's `.pack` files. And the `.git/index` file, if it is still memory-mapped. And when the spawned process is not quitting quickly, the *spawned* process can prevent those files from being renamed, modified or deleted.

The new code tries to ensure that only stdin/stdout/stderr are inherited by child processes.

So the code really tries to address a long-standing bug.

The problem with the new code is that it failed to anticipate *so many* issues with specifying those handles:
- On Windows 7 and older, a good deal of file handle types are simply disallowed in that list but are inherited *always, anyway* (#1475).
- And apparently `CreateProcess()` has code to be gentle with invalid standard handles being specified in `STARTUPINFO`, and that gentle treatment apparently does *not* extend to the code handling the `PROC_THREAD_ATTRIBUTE_HANDLE_LIST`.

These problems are really unfortunate, and quite honestly: unexpected. Nothing in the documentation prepared me for the hot-fixes required to patch up the issues.

So that's where we are. The new code intends to do the right thing, but needs to be adjusted to be a lot more fault-tolerant. And y'all are helping me greatly by being so responsive and testing and being patient. Thanks!

**Comment by dscho** [comment]
> I can show you the responsible code

Of course I should also link to the part above, that constructs the list of `stdhandles`: https://github.com/git-for-windows/git/blob/dc364ab06b41bcd142295751cef12e2c88fdc437/compat/mingw.c#L1655-L1666

**Comment by dscho** [comment]
I just uploaded a third custom installer: https://github.com/dscho/git/releases/tag/attempt-to-work-around-1481-v3

@MagicAndre1981 it should include `.pdb` files for Git's executables, could you try to work your magic?

To continue my musing about a proper fix (which I attempted via https://github.com/dscho/git/commit/868c7ceca349efb1b94ac21627a736f8bfecb833, but I am no longer certain that that would be the best work-around): Currently, the latest patch tries to exclude not only `INVALID_HANDLE_VALUE` but every handle that maps to a negative value.

I could imagine that a better fix may be to patch `winansi_get_osfhandle()` function to special-case `(HANDLE)-2` being returned from `_get_osfhandle()` and return `INVALID_HANDLE_VALUE` instead: https://github.com/git-for-windows/git/blob/dc364ab06b41bcd142295751cef12e2c88fdc437/compat/winansi.c#L657-L669

This would provide an alternate fix to https://github.com/dscho/git/commit/868c7ceca349efb1b94ac21627a736f8bfecb833 because `winansi_get_osfhandle()` is used to set the standard handles in `STARTUPINFO`: https://github.com/git-for-windows/git/blob/dc364ab06b41bcd142295751cef12e2c88fdc437/compat/mingw.c#L1651-L1653

**Comment by dscho** [comment]
Oh, and @TCHessen @0x084E if you got a minute, could you see whether the output with that third custom installer under `core.debug1481 = true` still mentions `handle #0: fffffffffffffffe` or another `handle #0` (it should not)?

**Comment by 0x084E** [comment]
Seems that there's one left:
```
warning: get_osf_handle for 7 (fd_is_interactive 0 0): 0000000000000284 (0000000000000000 0000000000000000)
warning: fhin: 7, hstdInput: 0000000000000284
warning: get_osf_handle for 8 (fd_is_interactive 0 0): 0000000000000288 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 000000000000063c (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 5
warning: get_osf_handle for 9 (fd_is_interactive 0 0): 000000000000015c (0000000000000000 0000000000000000)
warning: fhin: 9, hstdInput: 000000000000015c
warning: get_osf_handle for 10 (fd_is_interactive 0 0): 0000000000000160 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 000000000000063c (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 0
warning: get_osf_handle for 10 (fd_is_interactive 0 0): 000000000000018c (0000000000000000 0000000000000000)
warning: fhin: 10, hstdInput: 000000000000018c
warning: get_osf_handle for 12 (fd_is_interactive 0 0): 0000000000000194 (0000000000000000 0000000000000000)
warning: get_osf_handle for 11 (fd_is_interactive 0 0): 0000000000000190 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 0
warning: get_osf_handle for 5 (fd_is_interactive 0 0): 000000000000008c (0000000000000000 0000000000000000)
warning: fhin: 5, hstdInput: 000000000000008c
warning: get_osf_handle for 6 (fd_is_interactive 0 0): 0000000000000278 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 000000000000063c (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 5
warning: get_osf_handle for 11 (fd_is_interactive 0 0): 0000000000000190 (0000000000000000 0000000000000000)
warning: fhin: 11, hstdInput: 0000000000000190
warning: get_osf_handle for 12 (fd_is_interactive 0 0): 0000000000000194 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 000000000000063c (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 0
= [up to date]          develop            -> origin/develop
warning: get_osf_handle for 0 (fd_is_interactive 0 0): fffffffffffffffe (0000000000000000 0000000000000000)
warning: fhin: 0, hstdInput: fffffffffffffffe
warning: get_osf_handle for 1 (fd_is_interactive 0 0): 000000000000063c (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 000000000000063c (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 1, err: 0
```

**Comment by TCHessen** [comment]
```
git.exe pull --progress -v --no-rebase "origin"

warning: get_osf_handle for 7 (fd_is_interactive 0 0): 0000000000000280 (0000000000000000 0000000000000000)
warning: fhin: 7, hstdInput: 0000000000000280
warning: get_osf_handle for 8 (fd_is_interactive 0 0): 0000000000000284 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 00000000000005d8 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 5 (fd_is_interactive 0 0): 00000000000002a8 (0000000000000000 0000000000000000)
warning: fhin: 5, hstdInput: 00000000000002a8
warning: get_osf_handle for 6 (fd_is_interactive 0 0): 00000000000002ac (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 00000000000005d8 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 7 (fd_is_interactive 0 0): 000000000000014c (0000000000000000 0000000000000000)
warning: fhin: 7, hstdInput: 000000000000014c
warning: get_osf_handle for 8 (fd_is_interactive 0 0): 0000000000000150 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 00000000000005d8 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 8 (fd_is_interactive 0 0): 000000000000014c (0000000000000000 0000000000000000)
warning: fhin: 8, hstdInput: 000000000000014c
warning: get_osf_handle for 10 (fd_is_interactive 0 0): 0000000000000160 (0000000000000000 0000000000000000)
warning: get_osf_handle for 9 (fd_is_interactive 0 0): 0000000000000154 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 5 (fd_is_interactive 0 0): 00000000000002b4 (0000000000000000 0000000000000000)
warning: fhin: 5, hstdInput: 00000000000002b4
warning: get_osf_handle for 6 (fd_is_interactive 0 0): 00000000000002b8 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 00000000000005d8 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 9 (fd_is_interactive 0 0): 000000000000016c (0000000000000000 0000000000000000)
warning: fhin: 9, hstdInput: 000000000000016c
warning: get_osf_handle for 10 (fd_is_interactive 0 0): 000000000000013c (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 00000000000005d8 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
From https://git.yyy.local/Web/xxx
= [up to date]      master     -> origin/master
warning: get_osf_handle for 0 (fd_is_interactive 0 0): fffffffffffffffe (0000000000000000 0000000000000000)
warning: fhin: 0, hstdInput: fffffffffffffffe
warning: get_osf_handle for 1 (fd_is_interactive 0 0): 00000000000005d8 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 00000000000005d8 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 1, err: 122
warning: get_osf_handle for 0 (fd_is_interactive 0 0): fffffffffffffffe (0000000000000000 0000000000000000)
warning: fhin: 0, hstdInput: fffffffffffffffe
warning: get_osf_handle for 1 (fd_is_interactive 0 0): 00000000000005d8 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 00000000000005d8 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 1, err: 122
Already up to date.
warning: get_osf_handle for 0 (fd_is_interactive 0 0): fffffffffffffffe (0000000000000000 0000000000000000)
warning: fhin: 0, hstdInput: fffffffffffffffe
warning: get_osf_handle for 1 (fd_is_interactive 0 0): 00000000000005d8 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 00000000000005d8 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 1, err: 122

Success (1719 ms @ 07.02.2018 14:44:16)
```

**Comment by TCHessen** [comment]
@dscho I wrote an email, maybe we found a faster way to check it out.

**Comment by dscho** [comment]
@0x084E @TCHessen those outputs show success! `_get_osfhandle()` still returns `(HANDLE)-2`, which was kinda expected, but the code to fall back to calling `CreateProcess()` *without* restricting handle inheritance was not needed!

> I wrote an email, maybe we found a faster way to check it out.

@TCHessen let me check my mail ;-)

**Comment by willhausman** [comment]
I see that most people are reporting this error from TortoiseGit, with cmd.exe working correctly.  I am currently running git version 2.16.1.windows.2 from powershell and see the error on a simple pull, no extra args. Running Windows 7 Pro, version 6.1.

**Comment by dscho** [comment]
> I could imagine that a better fix may be to patch winansi_get_osfhandle() function to special-case (HANDLE)-2 being returned from _get_osfhandle() and return INVALID_HANDLE_VALUE instead

And here is an installer for a version that does precisely that: https://github.com/dscho/git/releases/tag/attempt-to-work-around-1481-v4

@0x084E @TCHessen could I ask you to test once again, maybe both with and without `core.debug1481` set?


> I am currently running git version 2.16.1.windows.2 from powershell and see the error on a simple pull, no extra args. Running Windows 7 Pro, version 6.1.

@willhausman you are not on the latest Git for Windows release. And based on your Windows version, I would expect you to hit another issue, reported as #1475: `cannot spawn: ... Function not implemented` (whereas the bug we are discussing here is about `cannot spawn: ... Invalid argument`).

I would expect Git for Windows v2.16.1(3) to have addressed your problem.

If not, please have a look at the custom installer I just posted.

**Comment by MagicAndre1981** [comment]
@dscho 

Here is output of v4:
```
git.exe pull --progress -v --no-rebase "origin"

warning: get_osf_handle for 8 (fd_is_interactive 0 0): 0000000000000140 (0000000000000000 0000000000000000)
warning: fhin: 8, hstdInput: 0000000000000140
warning: get_osf_handle for 9 (fd_is_interactive 0 0): 0000000000000144 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 0000000000000578 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 9 (fd_is_interactive 0 0): 0000000000000140 (0000000000000000 0000000000000000)
warning: fhin: 9, hstdInput: 0000000000000140
warning: get_osf_handle for 11 (fd_is_interactive 0 0): 0000000000000154 (0000000000000000 0000000000000000)
warning: get_osf_handle for 10 (fd_is_interactive 0 0): 0000000000000148 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 10 (fd_is_interactive 0 0): 0000000000000104 (0000000000000000 0000000000000000)
warning: fhin: 10, hstdInput: 0000000000000104
warning: get_osf_handle for 11 (fd_is_interactive 0 0): 0000000000000148 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 0000000000000578 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
From https://github.com/telerik/xamarin-sdk
= [up to date]      master     -> origin/master
warning: get_osf_handle for 0 (fd_is_interactive 0 0): fffffffffffffffe (0000000000000000 0000000000000000)
warning: fhin: 0, hstdInput: ffffffffffffffff
warning: get_osf_handle for 1 (fd_is_interactive 0 0): 0000000000000578 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 0000000000000578 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 1, err: 122
warning: get_osf_handle for 0 (fd_is_interactive 0 0): fffffffffffffffe (0000000000000000 0000000000000000)
warning: fhin: 0, hstdInput: ffffffffffffffff
warning: get_osf_handle for 1 (fd_is_interactive 0 0): 0000000000000578 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 0000000000000578 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 1, err: 122
Already up to date.
warning: get_osf_handle for 0 (fd_is_interactive 0 0): fffffffffffffffe (0000000000000000 0000000000000000)
warning: fhin: 0, hstdInput: ffffffffffffffff
warning: get_osf_handle for 1 (fd_is_interactive 0 0): 0000000000000578 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 0000000000000578 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 1, err: 122

Success (1063 ms @ 07.02.2018 15:14:56)
```

**Comment by TCHessen** [comment]
yeah... 

```
git.exe pull --progress -v --no-rebase "origin"

From https://git.yyy.local/Web/xxx
= [up to date]      master     -> origin/master
Already up to date.

Success (1453 ms @ 07.02.2018 15:15:13)

```

```
git.exe pull --progress -v --no-rebase "origin"

warning: get_osf_handle for 7 (fd_is_interactive 0 0): 000000000000028c (0000000000000000 0000000000000000)
warning: fhin: 7, hstdInput: 000000000000028c
warning: get_osf_handle for 8 (fd_is_interactive 0 0): 0000000000000290 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 5 (fd_is_interactive 0 0): 0000000000000280 (0000000000000000 0000000000000000)
warning: fhin: 5, hstdInput: 0000000000000280
warning: get_osf_handle for 6 (fd_is_interactive 0 0): 0000000000000288 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 7 (fd_is_interactive 0 0): 000000000000013c (0000000000000000 0000000000000000)
warning: fhin: 7, hstdInput: 000000000000013c
warning: get_osf_handle for 8 (fd_is_interactive 0 0): 0000000000000140 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 8 (fd_is_interactive 0 0): 000000000000013c (0000000000000000 0000000000000000)
warning: fhin: 8, hstdInput: 000000000000013c
warning: get_osf_handle for 10 (fd_is_interactive 0 0): 0000000000000154 (0000000000000000 0000000000000000)
warning: get_osf_handle for 9 (fd_is_interactive 0 0): 0000000000000148 (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 5 (fd_is_interactive 0 0): 0000000000000284 (0000000000000000 0000000000000000)
warning: fhin: 5, hstdInput: 0000000000000284
warning: get_osf_handle for 6 (fd_is_interactive 0 0): 0000000000000288 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
warning: get_osf_handle for 9 (fd_is_interactive 0 0): 0000000000000170 (0000000000000000 0000000000000000)
warning: fhin: 9, hstdInput: 0000000000000170
warning: get_osf_handle for 10 (fd_is_interactive 0 0): 0000000000000174 (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 3, err: 122
From https://git.yyy.local/Web/xxx
= [up to date]      master     -> origin/master
warning: get_osf_handle for 0 (fd_is_interactive 0 0): fffffffffffffffe (0000000000000000 0000000000000000)
warning: fhin: 0, hstdInput: ffffffffffffffff
warning: get_osf_handle for 1 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 1, err: 122
warning: get_osf_handle for 0 (fd_is_interactive 0 0): fffffffffffffffe (0000000000000000 0000000000000000)
warning: fhin: 0, hstdInput: ffffffffffffffff
warning: get_osf_handle for 1 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 1, err: 122
Already up to date.
warning: get_osf_handle for 0 (fd_is_interactive 0 0): fffffffffffffffe (0000000000000000 0000000000000000)
warning: fhin: 0, hstdInput: ffffffffffffffff
warning: get_osf_handle for 1 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: get_osf_handle for 2 (fd_is_interactive 0 0): 000000000000053c (0000000000000000 0000000000000000)
warning: ret: 1, restrict 1, #count: 1, err: 122

Success (1765 ms @ 07.02.2018 15:14:45)

```

**Comment by willhausman** [comment]
@dscho Correct on both fronts, apologies for the confusion.  I had just installed Git 2 days ago and assumed it was still the latest version. Confirmed that 2.16.3 works from powershell.

**Comment by dscho** [comment]
@MagicAndre1981 @TCHessen thank you so much for testing!

@MagicAndre1981 did you have any luck with those .pdb files and ETL? I would really be curious to know where that handle value comes from, if only to exclude Git for Windows as the culprit.

FWIW my current plan is to cherry-pick to `master` the patch that was at the tip of v4, and then also copy-edit https://github.com/dscho/git/commit/081b5fd19c9a945faa90222be25891fab5018641 to handle `ERROR_NO_SYSTEM_RESOURCES` as before (i.e. silently), but fall back to spawn without restricting the handle inheritance also upon any other errors albeit complaining loudly about it.

I have been bitten twice in a row by hot-fix versions of Git for Windows displaying serious bugs *the day of the release*, so I think I will let things simmer a little bit more lest we find yet another important bug to fix. But the snapshot builds at https://wingit.blob.core.windows.net/files/index.html should be working again, so there will be a semi-official installer (and portable Git and MinGit) available soon (the snapshots are signed just like the official Git for Windows versions, so you should be fine in the meantime).

**Comment by MagicAndre1981** [comment]
@dscho I found the PDBs, but the ETW doesn't show error code/result for the `Windows Kernel/Object/CreateHandle` events:

`    <Event MSec= "15326,6211" PID="2140" PName=     "git" TID="7140" EventName="Object/CreateHandle" Object="0xFFFFB28439D11970" Handle="0xFFFFFFFF80002718" ObjectType="36" ObjectName="" ObjectTypeName="File"/>`

But it would be nice if the official installer would also ship with PDBs becasue I already saw users having high CPU usage for git.exe. Here you can also use [ETW to see which functions cause high CPU usage](https://channel9.msdn.com/Shows/Defrag-Tools/Defrag-Tools-42-WPT-CPU-Analysis).

**Comment by MagicAndre1981** [comment]
@dscho 

I used the `Git-2.16.1-issue1481-4-64-bit.exe` now to do some pulls via TGit 2.5.7 and it works now fine for me.

**Comment by dscho** [comment]
> I found the PDBs, but the ETW doesn't show error code/result for the Windows Kernel/Object/CreateHandle events:
> 
> <Event MSec= "15326,6211" PID="2140" PName= "git" TID="7140" EventName="Object/CreateHandle" Object="0xFFFFB28439D11970" Handle="0xFFFFFFFF80002718" ObjectType="36" ObjectName="" ObjectTypeName="File"/>

@MagicAndre1981 thanks for doing this! I *guess* the limited output stems from the fact that those `.pdb` files were not generated by Visual Studio, but via `cv2pdb` from GCC-compiled code...

> But it would be nice if the official installer would also ship with PDBs

I planned on bundling a `.zip` file with the `.pdb` files for every Git for Windows version and offering it at the same place as the installers. I already have most of the building blocks in place, but started the final automation only yesterday, after publishing v2.16.1(3) (and I did not pick it up this morning because of this here ticket).

So yes, something like that is coming (not included in the installer, at least not at first, but as a separate `.zip` file).

> I used the Git-2.16.1-issue1481-4-64-bit.exe now to do some pulls via TGit 2.5.7 and it works now fine for me.

Excellent! Thanks for the feedback.

**Comment by MagicAndre1981** [comment]
@dscho no, the PDBs are only usefull to get the stack, otherwise you see ?!? in stack.  The missing error is a limitation of ETW. But having the PDBs now all time is great

**Comment by 0x084E** [comment]
@dscho ``MinGit-prerelease-2.16.0.windows.2.1027.g1a4ee4d5d8-64-bit`` is also working well, tried pull and stash with success.
Thanks for your work!

**Comment by dscho** [comment]
> MinGit-prerelease-2.16.0.windows.2.1027.g1a4ee4d5d8-64-bit is also working well

@0x084E that is unfortunately a stale snapshot... Ever since v2.16.1 was released, the snapshot builds failed (because I used Azure Blob's REST API incorrectly, and they started refusing incorrect usage a few days after that).

I will definitely notify y'all when a new snapshot is available. Or maybe I'll make up my mind and dare release v2.16.1(4), hoping that this time no crucial bug will be reported within hours of the release?

**Comment by MagicAndre1981** [comment]
@dscho 

release the v2.16.1.4 to fix this issue and unblock all TGit users.

**Comment by 0x084E** [comment]
Oops, I didn't notice this was an old release. It is getting late...

**Comment by dscho** [comment]
> release the v2.16.1.4 to fix this issue and unblock all TGit users.

@MagicAndre1981 @bkraul @0x084E okay, I'll do that.

> Oops, I didn't notice this was an old release. It is getting late...

@0x084E no worries.

**Comment by Hucklecatt** [comment]
Hi all - this is a blocker for our release. May I ask if (4) will be made available today (2/7)? Many thanks.

**Comment by dscho** [comment]
@Hucklecatt I was already on it, as I had said earlier. No need to put more stress on me.

**Comment by Hucklecatt** [comment]
My idiocy precedes me - really sorry about that Johannes.

**Comment by dscho** [comment]
@Hucklecatt no worries. I can see how you are eager to get a fixed version if your release (which software is that?) is blocked by the bug reported in this here ticket.

**Comment by dscho** [comment]
Voilà: https://github.com/git-for-windows/git/releases/tag/v2.16.1.windows.4

**Comment by 0x084E** [comment]
I confirm my previous confirmation - now on correct version -  that everything is working well again 😄 
Thanks!

**Comment by dakotahawkins** [comment]
@dscho I was having this issue generically, and it seems like there's still a problem:

```text
$ git checkout -
Switched to branch 'XXXX'
Your branch and 'origin/XXXX' have diverged,
and have 21 and 4 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)
warning: failed to restrict file handles (87)

handle #0: 0000000000000010 (type 2, handle info (1) 1d3a052
handle #1: 0000000000000098 (type 3, handle info (1) 1
handle #2: 000000000000001b (type 2, handle info (1) 1

This is a bug; please report it at
https://github.com/git-for-windows/git/issues/new

To suppress this warning, please set the environment variable

        SUPPRESS_HANDLE_INHERITANCE_WARNING=1

warning: failed to restrict file handles (87)

handle #0: 0000000000000010 (type 2, handle info (1) 1d3a052
handle #1: 0000000000000098 (type 3, handle info (1) 1
handle #2: 000000000000001f (type 2, handle info (1) 1

This is a bug; please report it at
https://github.com/git-for-windows/git/issues/new

To suppress this warning, please set the environment variable

        SUPPRESS_HANDLE_INHERITANCE_WARNING=1

warning: failed to restrict file handles (87)

handle #0: 0000000000000010 (type 2, handle info (1) 1d3a052
handle #1: 0000000000000098 (type 3, handle info (1) 1
handle #2: 000000000000001f (type 2, handle info (1) 1

This is a bug; please report it at
https://github.com/git-for-windows/git/issues/new

To suppress this warning, please set the environment variable

        SUPPRESS_HANDLE_INHERITANCE_WARNING=1

warning: failed to restrict file handles (87)

handle #0: 0000000000000010 (type 2, handle info (1) 1d34853
handle #1: 0000000000000270 (type 3, handle info (1) 1
handle #2: 000000000000001b (type 2, handle info (1) 1

This is a bug; please report it at
https://github.com/git-for-windows/git/issues/new

To suppress this warning, please set the environment variable

        SUPPRESS_HANDLE_INHERITANCE_WARNING=1
```

**Comment by bkraul** [comment]
Everything works here as well.

**Comment by dscho** [comment]
@dakotahawkins darn. At least you have a way to suppress the warning. 

FWIW `87` means `ERROR_INVALID_PARAMETER`. So there is *still* something about those handles that Windows 7 does not like (I am just going on a limb out here and assume that ConEmu on Windows 10 probably does not have the same problems).

I guess we should really try to suppress this warning in general on Windows 7 and earlier? Could you tell me the output of your `cmd //c ver` (in Bash)?

**Comment by dscho** [comment]
> > But it would be nice if the official installer would also ship with PDBs
> 
> I planned on bundling a .zip file with the .pdb files for every Git for Windows version and offering it at the same place as the installers. I already have most of the building blocks in place, but started the final automation only yesterday, after publishing v2.16.1(3) (and I did not pick it up this morning because of this here ticket).

@MagicAndre1981 I finished this work, and manually uploaded the files to the v2.16.1(4) release (they are called `pdbs-for-git-*.zip`). In the future, this should be automatic (modulo bugs I introduced ;-)).

**Comment by dakotahawkins** [comment]
@dscho Here you go: `Microsoft Windows [Version 6.1.7601]`

**Comment by csware** [comment]
@dscho 

>> If this flag is specified when calling the GetStartupInfo function, these members are either the handle value specified during process creation or INVALID_HANDLE_VALUE.
> 
> ... which suggests to me that hStdInput should either be set to a valid HANDLE or to INVALID_HANDLE_VALUE (which is -1, not 0). So maybe it comes from there?

I changed TortoiseGit to initialize `si.hStdInput = INVALID_HANDLE_VALUE;` still `warning: fhin: 0, hstdInput: fffffffffffffffe` is passed to `git.exe`...

If you have any idea what I can do on the TortoiseGit side, I'm happy to do.

**Comment by bkraul** [comment]
Ah...collaboration. A dream come true. Thanks guys for your work. You really keep the world going.

**Comment by dscho** [comment]
> I changed TortoiseGit to initialize si.hStdInput = INVALID_HANDLE_VALUE; still warning: fhin: 0, hstdInput: fffffffffffffffe is passed to git.exe...

That's funny, indeed... :-( Well, we do have a work-around for now.

**Comment by csware** [comment]
@dscho Even if we have a workaround now, if there's a bug in TortoiseGit I want to fix it...

**Comment by dscho** [comment]
> Here you go: Microsoft Windows [Version 6.1.7601]

@dakotahawkins I am currently thinking about something like this:

```diff
diff --git a/compat/mingw.c b/compat/mingw.c
index 89a516501e3..961f45ad3ad 100644
--- a/compat/mingw.c
+++ b/compat/mingw.c
@@ -1753,6 +1753,8 @@ static pid_t mingw_spawnve_fd(const char *cmd, const char **argv, char **deltaen
 	if (!ret && restrict_handle_inheritance && stdhandles_count) {
 		DWORD err = GetLastError();
 		if (err != ERROR_NO_SYSTEM_RESOURCES &&
+		    (err != ERROR_INVALID_PARAMETER ||
+		     GetVersion() >= 7601 << 16) &&
 		    !getenv("SUPPRESS_HANDLE_INHERITANCE_WARNING")) {
 			struct strbuf buf = STRBUF_INIT;
 			DWORD fl;
```

**Comment by dscho** [comment]
> Even if we have a workaround now, if there's a bug in TortoiseGit I want to fix it...

@csware since you changed the call, and since Git *still* reports an odd `HANDLE` value, I guess the culprit is not on TortoiseGit's side?

If you could dig further from here, that would be awesome. I really should go to bed to get well (been working on this despite being sick as I figured it's important).

**Comment by csware** [comment]
@dscho I'll do some more testing tomorrow (passing `dwFlags |= CREATE_NEW_CONSOLE` vs. `dwFlags |= DETACHED_PROCESS;` makes no difference).

I', also crashing down and "Gute Besserung".

**Comment by csware** [comment]
@dscho MSDN doesn't explecitly say that passing `INVALID_HANDLE_VALUE` is valid. I tested a bit with creating a pipe and closing the Write-Handle immediately after CreateProcess and this seems to work (git and ssh are using the ASKPASS helpers). This could be a better solution than passing `0` or ``INVALID_HANDLE_VALUE`, however, I'm unsure whether this has any negative side-effects (we had some issues with powershell in the past, I'm investigating further...).

**Comment by dakotahawkins** [comment]
@dscho The version check seems reasonable, but if I were you I'd consider putting the `7601 << 16` in a constant somewhere to make it more readable/obvious :)

**Comment by MagicAndre1981** [comment]
@dscho 

ok, I haven't checked the Git-2.16.1-issue1481-4-64-bit under Win7, only Win8.1 and Win10 1709 so I didn't noticed that it still has issues. Sorry that you still have trouble 🙈😇

**Comment by dscho** [comment]
@MagicAndre1981 no worries. I half expected it, which is why I introduced that environment variable that allows one to shut up the warning (but still forces the fall-back).

**Comment by dscho** [comment]
@dakotahawkins I finally committed a fix: 64b1c47b8100207e39432add27dbcab109f06763. Once the current snapshot is built (it will probably take another 20-30 minutes to complete), you should see it [here](https://wingit.blob.core.windows.net/files/index.html) (built from commit e1848984d1004040ec5199e749b5f282ddf4bb09). Can I ask you to test that one?
